File: EC EN 483/Practice_Final/Practice_Final/python/controllerObsv.py
import numpy as np
import rodMassParam as P

class controllerObsv:
    def __init__(self):
        self.observer_state = np.array([
            [0.0],  # estimate of theta
            [0.0],  # estimate of theta_hat
            [0.0],  # estimate of disturbance
        ])
        self.tau_d1 = 0.0  # control torque, delayed by one sample
        self.integrator = 0.0  # integrator
        self.error_d1 = 0.0  # error signal delayed by 1 sample
        self.K = P.K  # state feedback gain
        self.ki = P.ki2  # Input gain
        self.L = P.L2  # observer gain
        self.A = P.A2  # system model
        self.B = P.B2
        self.C = P.C2
        self.limit = P.tau_max
        self.tau_eq = P.tau_eq
        self.Ts = P.Ts  # sample rate of controller

    def update(self, theta_r, y):
        return tau, x_hat, d_hat

    def update_observer(self, y):
        # update the observer using RK4 integration
        F1 = self.observer_f(self.observer_state, y)
        F2 = self.observer_f(self.observer_state + self.Ts / 2 * F1, y)
        F3 = self.observer_f(self.observer_state + self.Ts / 2 * F2, y)
        F4 = self.observer_f(self.observer_state + self.Ts * F3, y)
        self.observer_state += self.Ts / 6 * (F1 + 2 * F2 + 2 * F3 + F4)
        x_hat = np.array([[self.observer_state.item(0)],[self.observer_state.item(1)]])
        d_hat = self.observer_state.item(2)
        return x_hat, d_hat

    def observer_f(self, x_hat, y):
        return xhat_dot

    def integrateError(self, error):

    def saturate(self,u):
        if abs(u) > self.limit:
            u = self.limit*np.sign(u)
        return u

File: EC EN 483/Practice_Final/Practice_Final/matlab/controllerObsv.m
classdef controllerObsv < handle
    properties
        obsv_state
        tau_d1
        integrator
        error_d1
        K
        ki
        L
        A
        B
        C
        limit
        Ts
        tau_eq
    end
    methods
        %--------constructor--------------------
        function self = controllerObsv(P)
            self.obsv_state = [0.0; 0.0; 0.0];
            self.tau_d1 = 0.0;
            self.integrator = 0.0;
            self.error_d1 = 0.0;
            self.K = P.K;
            self.ki = P.ki2;
            self.A = P.A2;
            self.B  = P.B2;
            self.C  = P.C2;
            self.L  = P.L2;
            self.limit = P.tau_max;
            self.Ts = P.Ts;
            self.tau_eq = P.tau_eq;
        end
        function [tau, x_hat, d_hat] = update(self, theta_r, y)
        end
        function [xhat, dhat] = updateObserver(self, y)
            % update observer using RK4 integration
            F1 = self.obsv_f(self.obsv_state, y);
            F2 = self.obsv_f(self.obsv_state + self.Ts/2*F1, y);
            F3 = self.obsv_f(self.obsv_state + self.Ts/2*F2, y);
            F4 = self.obsv_f(self.obsv_state + self.Ts*F3, y);
            self.obsv_state = self.obsv_state ...
                + self.Ts/6 * (F1 + 2*F2 + 2*F3 + F4);
            xhat = self.obsv_state(1:2);
            dhat = self.obsv_state(3);
        end
        function x_hat_dot = obsv_f(self, obsv_state, y)
        end
        function self = integrateError(self, error)
        end
        function out = saturate(self,u)
            if abs(u) > self.limit
                u = self.limit*sign(u);
            end
            out = u;
        end        
    end
end